### PR TITLE
fix(sbt-plugin): add new default registry URL

### PR DESCRIPTION
**Pull request by george-wilson-rea**
## Changes

Change the default repository for SBT plugin lookup to `https://repo1.maven.org/maven2/`

## Context

The bintray service has been sunset and now `https://repo.scala-sbt.org/scalasbt/` redirects to a mirror that no longer receives updates. Renovate should just default to maven.org for sbt plugins as it does for sbt packages. See this page for additional context: https://www.scala-sbt.org/1.x/docs/Bintray-For-Plugins.html

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/datasource/sbt-plugin/index.ts
Comment (viceice) [inline lib/modules/datasource/sbt-plugin/index.ts:12]: how do you verified that all existing versions are migrated to maven.org?

**Review by viceice — CHANGES_REQUESTED** [review]
you need to verify that it still works with older plugins.

File: lib/modules/datasource/sbt-plugin/index.ts
Comment (HonkingGoose) [inline lib/modules/datasource/sbt-plugin/index.ts:9]: > Change the default repository for SBT plugin lookup to https://repo1.maven.org/maven2/

**Maintainers**: is this a `fix:` or `feat:` type change. It feels more like a `feat:` or maybe even a breaking feature change?

Unless the entire thing is broken right now because everybody migrated to the new stuff and Renovate is still using the old defaults. In that case it's probably a `fix:` type change.

Comment (george-wilson-rea) [reply lib/modules/datasource/sbt-plugin/index.ts:9]: I believe it's a `fix`. We run self-hosted Renovate and we find that sbt plugins have not been updated for some time. Recently we configured our `sbt-plugins` to use `https://repo1.maven.org/maven2/` and now we are getting some plugin updates.

According to [this page](https://www.scala-sbt.org/1.x/docs/Bintray-For-Plugins.html) (emphasis mine):

>As of 2021-04-18, we have migrated all sbt plugins and sbt 0.13 artifacts to the Artifactory instance, and redirected https://repo.scala-sbt.org/scalasbt/ to point to it as well, so existing builds should continue to work without making any changes today and after May 1st. **For plugin hosting, we will operate this as a read-only repository. Any new plugin releases should migrate to using [Sonatype OSS](https://www.scala-sbt.org/1.x/docs/Using-Sonatype.html)**.

So I think Renovate has not been updating sbt plugins since 2021, because it's been looking at a read-only mirror that was frozen in 2021 (`https://repo.scala-sbt.org/scalasbt/`).

When I go to Sonatype OSS, their releases repo `https://oss.sonatype.org/content/repositories/releases/` just redirects to `https://repo1.maven.org/maven2/`, so that seems like the right thing to configure.

File: lib/modules/datasource/sbt-plugin/index.ts
Comment (george-wilson-rea) [reply lib/modules/datasource/sbt-plugin/index.ts]: Good point. I did not notice that older plugins aren't present in the newer repo. I've put the old URL back in, but I've made the maven URL the first in the list.

**Review by viceice — DISMISSED** [review]

**Review by HonkingGoose — COMMENTED** [review]
Did you check if there are readme files related to this change that need to be updated?

**Review by rarkins — CHANGES_REQUESTED** [review]
Needs at least one test on a "real" repository to make sure it works end to end. This is the type of change which would pass all unit tests while stopping plugin updating working altogether if something were wrong

**Comment by george-wilson-rea** [comment]
@HonkingGoose When I search for "`sbt`" in the `docs` directory, I don't see anything that would need updating. Should I be looking anywhere else?

@rarkins How should I test this on a "real" repository? Could you please elaborate?

**Comment by rarkins** [comment]
A public repo on GitHub.com which needs updates affected by this PR.

**Comment by george-wilson-rea** [comment]
Here's a repo: https://github.com/george-wilson-rea/renovate-scala-plugins

In `plugins.sbt`, `sbt-scoverage` is set to `2.0.11`, which is not the latest version. I expect it to be updated to `2.0.12` but with recommended config, that doesn't happen.

**Comment by george-wilson-rea** [comment]
Here's another repo, with an overridden sbt plugin config. Interestingly, sbt-scoverage is still not getting updated. I'm not sure what's going wrong here.
https://github.com/george-wilson-rea/renovate-scala-plugins-fixed

**Comment by george-wilson-rea** [comment]
Interestingly, it looks like sbt plugins _are_ working by default, but for some reason `sbt-scoverage` specifically doesn't work with either setup. I might have misdiagnosed the problem here.
How can I investigate what's going wrong with `sbt-scoverage`?

**Comment by HonkingGoose** [comment]
> @HonkingGoose When I search for "`sbt`" in the `docs` directory, I don't see anything that would need updating. Should I be looking anywhere else?

Hmm, the convention is that each manager/datasource/platform has a readme file. Maybe the sbt-plugin has no readme yet. It should have a readme though!

Please search in a similar directory for examples. 😉

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
Please add a readme to both the datasource and manager.

**Comment by george-wilson-rea** [comment]
@HonkingGoose The sbt manager has a readme saying
```
Extracts Maven artifact dependencies for SBT.
```
which is accurate. I have added some minimal readmes to the `sbt-package` and `sbt-plugin` datasources. Is there anything else I should say in these documents?

**Comment by george-wilson-rea** [comment]
I've started a discussion here with more detail about what seems to be going wrong: https://github.com/renovatebot/renovate/discussions/29755
I'm not sure this pull request will fix that problem but this pull request at least gives better defaults (new repo first, outdated repo second).

File: lib/modules/datasource/sbt-package/readme.md
Comment (HonkingGoose) [inline lib/modules/datasource/sbt-package/readme.md:1]:
```suggestion
This datasource finds SBT package updates from Maven repositories.
```

File: lib/modules/datasource/sbt-plugin/readme.md
Comment (HonkingGoose) [inline lib/modules/datasource/sbt-plugin/readme.md:1]:
```suggestion
This datasource finds SBT plugin updates from Maven repositories.
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
> ... I have added some minimal readmes to the `sbt-package` and `sbt-plugin` datasources. Is there anything else I should say in these documents?

It's usually good to mention the default settings in the readmes. So things like:

- By default, Renovate checks at human-readable-url for updates, but you can use `configOption=setting` to change this.
- By default, Renovate only finds updates when these conditions are met: list conditions.

File: lib/modules/datasource/sbt-plugin/index.spec.ts
Comment (HonkingGoose) [inline lib/modules/datasource/sbt-plugin/index.spec.ts:36]: > I've started a discussion here with more detail about what seems to be going wrong: #29755
> I'm not sure this pull request will fix that problem but this pull request at least gives better defaults (new repo first, outdated repo second).

I'm marking this PR as `status:blocked`. Once we are ready to merge, a maintainer should:

- [ ] Remove the `status:blocked` label
- [ ] Resolve this review comment

**Comment by george-wilson-rea** [comment]
I've now fixed my replication example and updated my bug report accordingly https://github.com/renovatebot/renovate/discussions/29755
This demonstrates that Renovate is not picking up plugin updates unless configured with extra registry URLs.

File: lib/modules/datasource/sbt-plugin/index.ts
Comment (zharinov) [inline lib/modules/datasource/sbt-plugin/index.ts:12]: The `MAVEN_REPO` constant from the `maven` datasource should be used instead

**Review by zharinov — CHANGES_REQUESTED** [review]

**Comment by zharinov** [comment]
We have to take into account different situations:
- Maven Central contains ancient plugin versions and `jfrog.io` contains the newer ones
- `jfrog.io` contains older versions, but newest ones are available in Maven Central

To address that, the `merge` registry strategy should be enabled.

File: lib/modules/datasource/sbt-plugin/index.ts
Comment (viceice) [reply lib/modules/datasource/sbt-plugin/index.ts:12]: good catch

**Review by viceice — COMMENTED** [review]
see comment

**Comment by zharinov** [comment]
Just opened PR with my suggestions: https://github.com/george-wilson-rea/renovate/pull/1

Because of `merge` registry strategy being used, now the exact order of registries doesn't really matter

File: lib/modules/datasource/sbt-plugin/index.ts
Comment (george-wilson-rea) [reply lib/modules/datasource/sbt-plugin/index.ts]: I've merged @zharinov 's pull request to address this.

File: lib/modules/datasource/sbt-package/index.ts
Comment (george-wilson-rea) [inline lib/modules/datasource/sbt-package/index.ts:29]: Now we've changed `sbt-plugin` to use the `merge` strategy - should `sbt-package` still be using `hunt` or should we change it to use `merge` too?

**Comment by george-wilson-rea** [comment]
I've merged the PR from @zharinov and fixed up the tests. This should be unblocked now.

File: lib/modules/datasource/sbt-package/index.ts
Comment (zharinov) [reply lib/modules/datasource/sbt-package/index.ts:29]: No, let's keep it as is, it should be fine

**Review by rarkins — DISMISSED** [review]

**Comment by zharinov** [comment]
Now we don't need test changes, check it out:
- https://github.com/george-wilson-rea/renovate/pull/2

**Comment by george-wilson-rea** [comment]
I've merged the latest suggestions from @zharinov, which has unfortunately dismissed the approving review. I've re-requested review per the contribution guidelines. Sorry for the churn.

**Review by zharinov — APPROVED** [review]

File: lib/modules/datasource/sbt-package/readme.md
Comment (HonkingGoose) [inline lib/modules/datasource/sbt-package/readme.md:3]:
```suggestion
By default, Renovate checks `https://repo.maven.apache.org/maven2` for SBT packages.
You can override the default behavior with the `registryUrls` config option.
For example:
```

File: lib/modules/datasource/sbt-plugin/readme.md
Comment (HonkingGoose) [inline lib/modules/datasource/sbt-plugin/readme.md:3]:
```suggestion
By default, Renovate:

1. Checks `https://repo1.maven.org/maven2/` for SBT plugins
1. If the above URL returns no results, then Renovate tries the _legacy_ URL: `https://repo.scala-sbt.org/scalasbt/sbt-plugin-releases`
 
You can override the default behavior with the `registryUrls` config option.
For example:
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
Some docs changes:

- Follow one sentence per line pattern
- Make a ordered list of the steps Renovate takes by default
- Small style fixes

**Comment by george-wilson-rea** [comment]
Thanks for enabling `auto-merge` @rarkins . Could you, or someone else, please also approve so that this gets merged? Your last review was dismissed when I merged @zharinov 's suggestions.

File: lib/modules/datasource/sbt-plugin/index.spec.ts
Comment (viceice) [inline lib/modules/datasource/sbt-plugin/index.spec.ts:41]:
```suggestion
          html`
```

use `html` template 😉 

https://www.npmjs.com/package/common-tags#html

File: lib/modules/datasource/sbt-plugin/index.spec.ts
Comment (viceice) [inline lib/modules/datasource/sbt-plugin/index.spec.ts:62]:
```suggestion
          html`
```

File: lib/modules/datasource/sbt-plugin/index.spec.ts
Comment (viceice) [inline lib/modules/datasource/sbt-plugin/index.spec.ts:77]:
```suggestion
          html`
```

File: lib/modules/datasource/sbt-plugin/index.spec.ts
Comment (viceice) [inline lib/modules/datasource/sbt-plugin/index.spec.ts:92]:
```suggestion
          html`
```

File: lib/modules/datasource/sbt-plugin/index.spec.ts
Comment (viceice) [inline lib/modules/datasource/sbt-plugin/index.spec.ts:108]:
```suggestion
          html`
```

File: lib/modules/datasource/sbt-plugin/index.spec.ts
Comment (viceice) [inline lib/modules/datasource/sbt-plugin/index.spec.ts:120]:
```suggestion
          html`
```

File: lib/modules/datasource/sbt-plugin/index.spec.ts
Comment (viceice) [inline lib/modules/datasource/sbt-plugin/index.spec.ts:1]:
```suggestion
import { codeBlock, html } from 'common-tags';
```

**Review by viceice — COMMENTED** [review]
otherwise LGTM

**Review by viceice — APPROVED** [review]

**Pull request merged**
